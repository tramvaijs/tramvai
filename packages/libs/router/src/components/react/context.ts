import { createContext } from 'react';
import type { Url } from '@tinkoff/url';
import type { AbstractRouter } from '../../router/abstract';
import type { NavigationRoute } from '../../types';

export const RouterContext = createContext<AbstractRouter>(null);
export const RouteContext = createContext<NavigationRoute>(null);
export const UrlContext = createContext<Url>(null);

export type ViewTransitionState =
  | {
      isTransitioning: false;
    }
  | {
      isTransitioning: true;
      currentRoute: NavigationRoute;
      nextRoute: NavigationRoute;
    };

export const ViewTransitionContext = createContext<ViewTransitionState>({
  isTransitioning: false,
});
