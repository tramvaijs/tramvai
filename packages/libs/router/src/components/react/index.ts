import { TransitionsProvider } from './providers/transitions';
import { DefaultProvider } from './providers/default';

export type { RouterState } from './providers/types';
export { useRouter } from './useRouter';
export { useRoute } from './useRoute';
export { useUrl } from './useUrl';
export { useNavigate } from './useNavigate';
export { useViewTransition } from './useViewTransition';

export const Provider =
  process.env.__TRAMVAI_VIEW_TRANSITIONS === 'true' ||
  process.env.__TRAMVAI_REACT_TRANSITIONS === 'true'
    ? TransitionsProvider
    : DefaultProvider;
