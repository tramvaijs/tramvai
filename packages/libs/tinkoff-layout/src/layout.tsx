import React from 'react';
import type { ReactComponent, LayoutProps, LayoutOptions } from './types.h';
import { composeComponent } from './utils';

const RenderChildren: ReactComponent = ({ children }) => children;

const RenderProp: ReactComponent = ({ children, ...props }) => children(props);

export const createLayout = ({ components = {}, wrappers = {} }: LayoutOptions = {}) => {
  const {
    header: BaseHeader = RenderProp,
    footer: BaseFooter = RenderProp,
    layout: BaseLayout = RenderChildren,
    content: BaseContent = RenderChildren,
    page: BasePage = RenderChildren,
    ...globalComponents
  } = components;

  const LayoutWrapper = composeComponent(BaseLayout, wrappers.layout);
  const ContentWrapper = composeComponent(BaseContent, wrappers.content);
  const PageWrapper = composeComponent(BasePage, wrappers.page);
  const HeaderWrapper = composeComponent(BaseHeader, wrappers.header);
  const FooterWrapper = composeComponent(BaseFooter, wrappers.footer);

  const componentsList = Object.keys(globalComponents).map((name) => {
    const Component = composeComponent(globalComponents[name], wrappers[name]);
    return { name, Component };
  });

  const Layout: React.FC<LayoutProps> = ({ children, Header, Footer }) => {
    return (
      <LayoutWrapper>
        {componentsList.map(({ Component, name }) => (
          <Component key={name} />
        ))}
        <ContentWrapper>
          {Header ? <HeaderWrapper>{(props: any) => <Header {...props} />}</HeaderWrapper> : null}
          <PageWrapper>{children}</PageWrapper>
          {Footer ? <FooterWrapper>{(props: any) => <Footer {...props} />}</FooterWrapper> : null}
        </ContentWrapper>
      </LayoutWrapper>
    );
  };

  Layout.displayName = 'Layout';

  return Layout;
};
