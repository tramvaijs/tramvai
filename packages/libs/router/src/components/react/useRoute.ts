import { useContext } from 'react';
import { RouteContext } from './context';

export const useRoute = () => {
  return useContext(RouteContext);
};
