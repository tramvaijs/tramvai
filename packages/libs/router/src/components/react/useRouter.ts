import { useContext } from 'react';
import { RouterContext } from './context';

export const useRouter = () => {
  return useContext(RouterContext);
};
