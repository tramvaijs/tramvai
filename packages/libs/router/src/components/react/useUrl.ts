import { useContext } from 'react';
import { UrlContext } from './context';

export const useUrl = () => {
  return useContext(UrlContext);
};
