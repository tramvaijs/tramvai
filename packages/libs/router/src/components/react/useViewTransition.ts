import { useContext } from 'react';

import type { AbstractRouter } from '../../router/abstract';
import type { NavigateOptions } from '../../types';
import { ViewTransitionContext, RouterContext } from './context';

type Payload = string | Partial<NavigateOptions>;

const normalizePath = (payload: Payload, router: AbstractRouter): string => {
  const route = router.resolve(payload);
  const resolvedPath =
    route.redirect !== undefined ? router.resolve(route.redirect).actualPath : route.actualPath;

  return resolvedPath.endsWith('/') ? resolvedPath : `${resolvedPath}/`;
};

export const useViewTransition = (payload: Payload): boolean => {
  const viewTransition = useContext(ViewTransitionContext);
  const router = useContext(RouterContext);

  if (viewTransition.isTransitioning) {
    const path = normalizePath(payload, router);
    const currentPath = viewTransition.currentRoute.actualPath;
    const nextPath = viewTransition.nextRoute.actualPath;

    // We are handling forward and back navigations for the same route here
    return [currentPath, nextPath].includes(path);
  }

  return false;
};
