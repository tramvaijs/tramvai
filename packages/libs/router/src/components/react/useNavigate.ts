import { useCallback } from 'react';
import { useShallowEqual } from '@tinkoff/react-hooks';
import { useRouter } from './useRouter';
import type { NavigateOptions } from '../../types';

interface UseNavigate {
  (rootOptions: Partial<NavigateOptions> | string): () => Promise<void>;
  (): (specificOptions: NavigateOptions | string) => Promise<void>;
}

const convertToNavigateOptions = (options: Partial<NavigateOptions> | string) => {
  return typeof options === 'string' ? { url: options } : options;
};

export const useNavigate: UseNavigate = (rootOptions?: Partial<NavigateOptions> | string) => {
  const router = useRouter();
  const rootOpts = useShallowEqual(convertToNavigateOptions(rootOptions));

  return useCallback(
    (specificOptions?: NavigateOptions | string) => {
      const opts = rootOpts ?? convertToNavigateOptions(specificOptions);

      return router.navigate(opts);
    },
    [rootOpts, router]
  );
};
