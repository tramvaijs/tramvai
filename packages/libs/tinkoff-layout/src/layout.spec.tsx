/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { createLayout } from './layout';

const Page = () => <div>Page</div>;
const Header = ({ prop }) => <header>head{prop ? ` prop=${prop}` : null}</header>;
const Footer = ({ prop }) => <footer>foot{prop ? ` prop=${prop}` : null}</footer>;

describe('createLayout', () => {
  it('layout with empty parameters', () => {
    const Layout = createLayout({
      wrappers: {},
      components: {},
    });
    const { container } = render(
      <Layout>
        <Page />
      </Layout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('layout with nothing', () => {
    const Layout = createLayout();
    const { container } = render(
      <Layout>
        <Page />
      </Layout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('layout with full options', () => {
    const Layout = createLayout({
      wrappers: {
        footer: [
          (Wrapped) => (props) =>
            (
              <div>
                Footer
                <Wrapped {...props} />
              </div>
            ),
        ],
        header: [
          (Wrapped) => (props) =>
            (
              <div>
                Header
                <Wrapped {...props} />
              </div>
            ),
          (Wrapped) => (props) =>
            (
              <div className="header-wrapper">
                <Wrapped {...props} />
              </div>
            ),
        ],
        page: [
          (Wrapped) => (props) =>
            (
              <div className="page-wrapper">
                <Wrapped {...props} />
              </div>
            ),
        ],
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
        content: [
          (Wrapped) => (props) =>
            (
              <div className="page-wrapper">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
      components: {
        global1: () => <div>global component</div>,
        global2: () => <div>global2 component</div>,
      },
    });

    const { container } = render(
      <Layout Header={Header} Footer={Footer}>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('layout with only layout wrapper', () => {
    const Layout = createLayout({
      components: {
        feedback: () => <div>fedback</div>,
        popup: () => <div>popup</div>,
      },
      wrappers: {
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
    });

    const { container } = render(
      <Layout>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('layout with header and footer', () => {
    const Layout = createLayout({
      components: {
        feedback: () => <div>fedback</div>,
        popup: () => <div>popup</div>,
      },
      wrappers: {
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
    });

    const { container } = render(
      <Layout Header={Header} Footer={Footer}>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('layout with base layout, content and page components', () => {
    const Layout = createLayout({
      wrappers: {
        layout: (Wrapped) => (props) => <Wrapped {...props} className="layout" />,
        content: (Wrapped) => (props) => <Wrapped {...props} className="content" />,
        page: (Wrapped) => (props) => <Wrapped {...props} className="page" />,
      },
      components: {
        layout: ({ children, className }) => (
          <div className={className}>
            Base layout
            {children}
          </div>
        ),
        content: ({ children, className }) => (
          <div className={className}>
            Content page
            {children}
          </div>
        ),
        page: ({ children, className }) => (
          <div className={className}>
            Base page
            {children}
          </div>
        ),
      },
    });

    const { container } = render(
      <Layout>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render wrappers when Header and Footer is missing', () => {
    const Layout = createLayout({
      wrappers: {
        header: [
          (Wrapped) => (props) =>
            (
              <div className="header">
                <Wrapped {...props} />
              </div>
            ),
        ],
        footer: [
          (Wrapped) => (props) =>
            (
              <div className="footer">
                <Wrapped {...props} />
              </div>
            ),
        ],
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
    });

    const { container } = render(
      <Layout>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not render Header when some of wrappers return null', () => {
    const Layout = createLayout({
      wrappers: {
        header: [
          (Wrapped) => (props) =>
            (
              <div className="header">
                <Wrapped {...props} />
              </div>
            ),
          (Wrapped) => (props) => null,
        ],
        footer: [
          (Wrapped) => (props) =>
            (
              <div className="footer">
                <Wrapped {...props} />
              </div>
            ),
        ],
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
    });

    const { container } = render(
      <Layout Header={Header} Footer={Footer}>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should pass custom props from some of wrappers to Header and Footer', () => {
    const Layout = createLayout({
      wrappers: {
        header: [
          (Wrapped) => (props) =>
            (
              <div className="header">
                <Wrapped {...props} />
              </div>
            ),
          (Wrapped) => (props) => <Wrapped {...props} prop="a" />,
        ],
        footer: [
          (Wrapped) => (props) =>
            (
              <div className="footer">
                <Wrapped {...props} prop="b" />
              </div>
            ),
        ],
        layout: [
          (Wrapped) => (props) =>
            (
              <div className="layout">
                <Wrapped {...props} />
              </div>
            ),
        ],
      },
    });

    const { container } = render(
      <Layout Header={Header} Footer={Footer}>
        <Page />
      </Layout>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
