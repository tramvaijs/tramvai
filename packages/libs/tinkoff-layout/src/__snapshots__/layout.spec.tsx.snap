// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLayout layout with base layout, content and page components 1`] = `
<div
  class="layout"
>
  Base layout
  <div
    class="content"
  >
    Content page
    <div
      class="page"
    >
      Base page
      <div>
        Page
      </div>
    </div>
  </div>
</div>
`;

exports[`createLayout layout with empty parameters 1`] = `
<div>
  Page
</div>
`;

exports[`createLayout layout with full options 1`] = `
<div
  class="layout"
>
  <div>
    global component
  </div>
  <div>
    global2 component
  </div>
  <div
    class="page-wrapper"
  >
    <div
      class="header-wrapper"
    >
      <div>
        Header
        <header>
          head
        </header>
      </div>
    </div>
    <div
      class="page-wrapper"
    >
      <div>
        Page
      </div>
    </div>
    <div>
      Footer
      <footer>
        foot
      </footer>
    </div>
  </div>
</div>
`;

exports[`createLayout layout with header and footer 1`] = `
<div
  class="layout"
>
  <div>
    fedback
  </div>
  <div>
    popup
  </div>
  <header>
    head
  </header>
  <div>
    Page
  </div>
  <footer>
    foot
  </footer>
</div>
`;

exports[`createLayout layout with nothing 1`] = `
<div>
  Page
</div>
`;

exports[`createLayout layout with only layout wrapper 1`] = `
<div
  class="layout"
>
  <div>
    fedback
  </div>
  <div>
    popup
  </div>
  <div>
    Page
  </div>
</div>
`;

exports[`createLayout should not render Header when some of wrappers return null 1`] = `
<div
  class="layout"
>
  <div>
    Page
  </div>
  <div
    class="footer"
  >
    <footer>
      foot
    </footer>
  </div>
</div>
`;

exports[`createLayout should not render wrappers when Header and Footer is missing 1`] = `
<div
  class="layout"
>
  <div>
    Page
  </div>
</div>
`;

exports[`createLayout should pass custom props from some of wrappers to Header and Footer 1`] = `
<div
  class="layout"
>
  <div
    class="header"
  >
    <header>
      head
       prop=a
    </header>
  </div>
  <div>
    Page
  </div>
  <div
    class="footer"
  >
    <footer>
      foot
       prop=b
    </footer>
  </div>
</div>
`;
